builtin/logical/ssh/path_role_create.go

```diff
@@ -0,0 +1,197 @@
+		return logical.ErrorResponse(fmt.Sprintf("Role '%s' not found", roleName)), nil
+	}
+	var role sshRole
+	if err := roleEntry.DecodeJSON(&role); err != nil {
+		return nil, err
+	}
+
+	if username == "" {
+		username = role.DefaultUser
+	}
+
+	//validate the IP address
+	ipAddr := net.ParseIP(ipRaw)
+	if ipAddr == nil {
+		return logical.ErrorResponse(fmt.Sprintf("Invalid IP '%s'", ipRaw)), nil
+	}
+	ip := ipAddr.String()
+
+	ipMatched := false
+	for _, item := range strings.Split(role.CIDR, ",") {
+		_, cidrIPNet, err := net.ParseCIDR(item)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Invalid cidr entry '%s'", item)), nil
+		}
+		ipMatched = cidrIPNet.Contains(ipAddr)
+		if ipMatched {
+			break
+		}
+	}
+	if !ipMatched {
+		return logical.ErrorResponse(fmt.Sprintf("IP[%s] does not belong to role[%s]", ip, roleName)), nil
+	}
+
+	//fetch the host key to be used for installation
+	keyPath := "keys/" + role.KeyName
+	keyEntry, err := req.Storage.Get(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("Key '%s' not found error:%s", role.KeyName, err)
+	}
+	var hostKey sshHostKey
+	if err := keyEntry.DecodeJSON(&hostKey); err != nil {
+		return nil, fmt.Errorf("Error reading the host key: %s", err)
+	}
+
+	//store the host key to file. Use it as parameter for scp command
+	hostKeyFileName := "./vault_ssh_" + username + "_" + ip + "_shared.pem"
+	err = ioutil.WriteFile(hostKeyFileName, []byte(hostKey.Key), 0600)
+
+	dynamicPrivateKeyFileName := "vault_ssh_" + username + "_" + ip + "_otk.pem"
+	dynamicPublicKeyFileName := dynamicPrivateKeyFileName + ".pub"
+
+	//delete the temporary files if they are already present
+	err = removeFile(dynamicPrivateKeyFileName)
+	if err != nil {
+		return nil, fmt.Errorf(fmt.Sprintf("Error removing dynamic private key file: '%s'", err))
+	}
+	err = removeFile(dynamicPublicKeyFileName)
+	if err != nil {
+		return nil, fmt.Errorf(fmt.Sprintf("Error removing dynamic private key file: '%s'", err))
+	}
+
+	//generate RSA key pair
+	dynamicPublicKey, dynamicPrivateKey, _ := generateRSAKeys()
+
+	//save the public key pair to a file
+	ioutil.WriteFile(dynamicPublicKeyFileName, []byte(dynamicPublicKey), 0644)
+
+	//send the public key to target machine
+	err = uploadFileScp(dynamicPublicKeyFileName, username, ip, hostKey.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	//connect to target machine
+	session, err := createSSHPublicKeysSession(username, ip, hostKey.Key)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create SSH Session using public keys: %s", err)
+	}
+	if session == nil {
+		return nil, fmt.Errorf("Invalid session object")
+	}
+	var buf bytes.Buffer
+	session.Stdout = &buf
+
+	authKeysFileName := "/home/" + username + "/.ssh/authorized_keys"
+	tempKeysFileName := "/home/" + username + "/temp_authorized_keys"
+
+	//commands to be run on target machine
+	grepCmd := "grep -vFf " + dynamicPublicKeyFileName + " " + authKeysFileName + " > " + tempKeysFileName + ";"
```

Comment:
I'll think about the idea you shared. (Deploying a binary to target machine to get the job done instead of depending on shell commands)